### PR TITLE
QT-4888: Implemented redislock using radix client

### PR DESCRIPTION
**Pull request by rajatkumar79**

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -6,9 +6,6 @@ test:
 test-race:
 	go test ./... -race
 
-lint:
-	golangci-lint run
-
```
Comment (ayush-singhal-2021) [inline Makefile:11]: can add option to run lint and fmt in Makefile itself.

File: README.md [hunk 2]
```diff
@@ -1,11 +1,7 @@
 # redislock
 
-[![Test](https://github.com/bsm/redislock/actions/workflows/test.yml/badge.svg)](https://github.com/bsm/redislock/actions/workflows/test.yml)
-[![GoDoc](https://godoc.org/github.com/bsm/redislock?status.png)](http://godoc.org/github.com/bsm/redislock)
-[![License](https://img.shields.io/badge/License-Apache%202.0-blue.svg)](https://opensource.org/licenses/Apache-2.0)
+Simplified distributed locking implementation using Redis with [radix](https://github.com/mediocregopher/radix) as client. For more information, please see examples.
```
Comment (ayush-singhal-2021) [inline README.md:4]: Add the link to original library too, of which this one is a fork

File: redislock.go [hunk 3]
```diff
@@ -145,44 +141,41 @@ func (l *Lock) Metadata() string {
...
+	var num int64
+	err := l.client.client.Do(luaPTTL.Cmd(&num, l.key, l.value))
+	if err != nil {
 		return 0, err
 	}
-
-	if num := res.(int64); num > 0 {
-		return time.Duration(num) * time.Millisecond, nil
+	if num < 0 {
+		return 0, nil
```
Comment (ayush-singhal-2021) [inline redislock.go]:
```suggestion
	if err != nil || num < 0 {
		return 0, err
```

File: redislock.go [hunk 4]
```diff
@@ -145,44 +141,41 @@ func (l *Lock) Metadata() string {
...
-	res, err := luaPTTL.Run(ctx, l.client.client, []string{l.key}, l.value).Result()
-	if err == redis.Nil {
-		return 0, nil
-	} else if err != nil {
+func (l *Lock) TTL() (time.Duration, error) {
+	var num int64
+	err := l.client.client.Do(luaPTTL.Cmd(&num, l.key, l.value))
+	if err != nil {
 		return 0, err
 	}
```
Comment (ayush-singhal-2021) [inline redislock.go]: this won't be required if you apply the suggestion

File: redislock.go [hunk 5]
```diff
@@ -145,44 +141,41 @@ func (l *Lock) Metadata() string {
...
+	return time.Duration(num) * time.Millisecond, nil
 }
 
 // Refresh extends the lock with a new TTL.
 // May return ErrNotObtained if refresh is unsuccessful.
-func (l *Lock) Refresh(ctx context.Context, ttl time.Duration, opt *Options) error {
-	ttlVal := strconv.FormatInt(int64(ttl/time.Millisecond), 10)
-	status, err := luaRefresh.Run(ctx, l.client.client, []string{l.key}, l.value, ttlVal).Result()
+func (l *Lock) Refresh(ttl time.Duration, opt *Options) error {
+	status := false
```
Comment (ayush-singhal-2021) [inline redislock.go]:
```suggestion
	var status bool
```

File: redislock.go [hunk 6]
```diff
@@ -145,44 +141,41 @@ func (l *Lock) Metadata() string {
...
+	err := l.client.client.Do(luaRefresh.Cmd(&status, l.key, l.value, strconv.FormatInt(ttl.Milliseconds(), 10)))
 	if err != nil {
```
Comment (ayush-singhal-2021) [inline redislock.go]:
```suggestion
	if err != nil || status {
```

File: redislock.go [hunk 7]
```diff
@@ -145,44 +141,41 @@ func (l *Lock) Metadata() string {
...
 		return err
-	} else if status == int64(1) {
-		return nil
 	}
-	return ErrNotObtained
+	if !status {
+		return ErrNotObtained
+	}
+	return nil
```
Comment (ayush-singhal-2021) [inline redislock.go]:
```suggestion
	return ErrNotObtained
```

File: redislock.go [hunk 8]
```diff
@@ -145,44 +141,41 @@ func (l *Lock) Metadata() string {
...
 
 // Release manually releases the lock.
 // May return ErrLockNotHeld.
-func (l *Lock) Release(ctx context.Context) error {
-	res, err := luaRelease.Run(ctx, l.client.client, []string{l.key}, l.value).Result()
-	if err == redis.Nil {
-		return ErrLockNotHeld
-	} else if err != nil {
+func (l *Lock) Release() error {
+	num := 0
```
Comment (ayush-singhal-2021) [inline redislock.go]:
```suggestion
	var num int
```

File: redislock.go [hunk 9]
```diff
@@ -28,49 +28,40 @@ var (
...
 	token, err := c.randomToken()
 	if err != nil {
 		return nil, err
 	}
 
 	value := token + opt.getMetadata()
+	ctx := opt.getContext()
 	retry := opt.getRetryStrategy()
 
-	// make sure we don't retry forever
```
Comment (ayush-singhal-2021) [inline redislock.go:64]: don't remove it, update the same instead

File: redislock.go [hunk 10]
```diff
@@ -28,49 +28,40 @@ var (
...
+	var cancel context.CancelFunc = func() {}
 	if _, ok := ctx.Deadline(); !ok {
-		var cancel context.CancelFunc
 		ctx, cancel = context.WithDeadline(ctx, time.Now().Add(ttl))
-		defer cancel()
```
Comment (ayush-singhal-2021) [inline redislock.go]: why did we change this logic ?

File: redislock.go [hunk 11]
```diff
@@ -97,8 +88,13 @@ func (c *Client) Obtain(ctx context.Context, key string, ttl time.Duration, opt
...
 
-func (c *Client) obtain(ctx context.Context, key, value string, ttl time.Duration) (bool, error) {
-	return c.client.SetNX(ctx, key, value, ttl).Result()
+func (c *Client) obtain(key, value string, ttl time.Duration) (bool, error) {
+	var result string
+	err := c.client.Do(radix.FlatCmd(&result, "SET", key, value, "PX", ttl.Milliseconds(), "NX"))
+	if err != nil {
+		return false, err
+	}
+	return result == "OK", nil
```
Comment (ayush-singhal-2021) [inline redislock.go]:
```suggestion
	return result == "OK", err
```

**Review by ayush-singhal-2021 — CHANGES_REQUESTED** [review]

File: redislock.go (hunk 7 above)
Comment (ayush-singhal-2021) [reply redislock.go]: please fix this

File: redislock_test.go [hunk 12]
```diff
@@ -1,273 +1,197 @@
...
-	}
-
-	// lock again
-	lock, err = client.Obtain(ctx, lockKey, time.Hour, nil)
-	if err != nil {
-		t.Fatal(err)
-	}
-	defer lock.Release(ctx)
-}
+var _ = Describe("Client", func() {
```
Comment (ayush-singhal-2021) [inline redislock_test.go:18]:
```suggestion
_ := Describe("Client", func() {
```

File: redislock_test.go [hunk 13]
```diff
@@ -1,273 +1,197 @@
...
+})
 
-	if err := rc.Del(context.Background(), lockKey).Err(); err != nil {
-		t.Fatal(err)
-	}
-	if err := rc.Close(); err != nil {
-		t.Fatal(err)
-	}
-}
+var _ = AfterSuite(func() {
```
Comment (ayush-singhal-2021) [inline redislock_test.go:195]: same, fix everywhere

File: redislock.go [hunk 14]
```diff
@@ -145,44 +139,35 @@ func (l *Lock) Metadata() string {
...
-	res, err := luaRelease.Run(ctx, l.client.client, []string{l.key}, l.value).Result()
-	if err == redis.Nil {
-		return ErrLockNotHeld
-	} else if err != nil {
+func (l *Lock) Release() error {
+	var num int
+	err := l.client.client.Do(luaRelease.Cmd(&num, l.key, l.value))
+	if err != nil {
 		return err
 	}
```
Comment (ayush-singhal-2021) [inline redislock.go]:
```suggestion
	if err != nil || num == 1 {
		return err
	}
```

File: redislock.go [hunk 15]
```diff
@@ -145,44 +139,35 @@ func (l *Lock) Metadata() string {
...
-
-	if i, ok := res.(int64); !ok || i != 1 {
+	if num != 1 {
 		return ErrLockNotHeld
 	}
 	return nil
```
Comment (ayush-singhal-2021) [inline redislock.go]:
```suggestion
	return ErrLockNotHeld
```

**Review by ayush-singhal-2021 — CHANGES_REQUESTED** [review]

File: redislock_test.go (hunk 12 above)
Comment (rajatkumar79) [reply redislock_test.go:18]: Such declaration is not allowed. Getting `expected declaration, found _`

File: redislock_test.go (hunk 13 above)
Comment (rajatkumar79) [reply redislock_test.go:195]: same

File: redislock.go [hunk 16]
```diff
@@ -145,59 +139,52 @@ func (l *Lock) Metadata() string {
...
 
 // TTL returns the remaining time-to-live. Returns 0 if the lock has expired.
-func (l *Lock) TTL(ctx context.Context) (time.Duration, error) {
-	res, err := luaPTTL.Run(ctx, l.client.client, []string{l.key}, l.value).Result()
-	if err == redis.Nil {
+func (l *Lock) TTL() (time.Duration, error) {
+	var num int64
+	err := l.client.client.Do(luaPTTL.Cmd(&num, l.key, l.value))
+	if err != nil || num < 0 {
 		return 0, nil
```
Comment (ayush-singhal-2021) [inline redislock.go]: forgot to return the error ?

**Review by ayush-singhal-2021 — CHANGES_REQUESTED** [review]

**Review by ayush-singhal-2021 — APPROVED** [review]

**Pull request merged**